service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
+	respChan    chan bridgeResponse
+	respWritten bool
+}
+
+func (w *requestResponseWriter) Header() *prot.MessageHeader {
+	return w.header
+}
+
+func (w *requestResponseWriter) Write(r interface{}) {
+	w.respChan <- bridgeResponse{header: w.header, response: r}
+	w.respWritten = true
+}
+
+func (w *requestResponseWriter) Error(err error) {
+	resp := &prot.MessageResponseBase{}
+	setErrorForResponseBase(resp, err)
+	w.Write(resp)
+}
+
+// Bridge defines the bridge client in the GCS. It acts in many
+// ways analogous to go's `http` package and multiplexer.
+//
+// It has two fundamentally different dispatch options:
+//
+// 1. Request/Response where by using the `Handler` a request
+//    of a given type will be dispatched to the apprpriate handler
+//    and an appropriate `ResponseWriter` will respond to exactly
+//    that request that caused the dispatch.
+//
+// 2. `PublishNotification` where by a notification that was not initiated
+//    by a request from any client can be written to the bridge at any time
+//    in any order.
+type Bridge struct {
+	// Transport is the transport interface used by the bridge.
+	Transport transport.Transport
+
+	// Handler to invoke when messages are received
+	Handler Handler
+
+	// commandConn is the Connection the bridge receives commands (such as
+	// ComputeSystemCreate) over.
+	commandConn transport.Connection
+
+	responseChan chan bridgeResponse
+
+	// Core - TODO: Remove this and use the mux!
+	coreint core.Core
+
+	// testing hook to close the bridge ListenAndServe() method
+	quitChan chan bool
+}
+
+// AssignHandlers creates and assigns the appropriate bridge
+// events to be listen for and intercepted on `mux` before forwarding
+// to `gcs` for handling.
+func (b *Bridge) AssignHandlers(mux *Mux, gcs core.Core) {
+	b.coreint = gcs
+	mux.HandleFunc(prot.ComputeSystemCreateV1, b.createContainer)
+	mux.HandleFunc(prot.ComputeSystemExecuteProcessV1, b.execProcess)
+	mux.HandleFunc(prot.ComputeSystemShutdownForcedV1, b.killContainer)
+	mux.HandleFunc(prot.ComputeSystemShutdownGracefulV1, b.shutdownContainer)
+	mux.HandleFunc(prot.ComputeSystemSignalProcessV1, b.signalProcess)
+	mux.HandleFunc(prot.ComputeSystemGetPropertiesV1, b.listProcesses)
+	mux.HandleFunc(prot.ComputeSystemWaitForProcessV1, b.waitOnProcess)
+	mux.HandleFunc(prot.ComputeSystemResizeConsoleV1, b.resizeConsole)
+	mux.HandleFunc(prot.ComputeSystemModifySettingsV1, b.modifySettings)
+}
+
+// ListenAndServe connects to the bridge transport, listens for
+// messages and dispatches the appropriate handlers to handle each
+// event in an asynchronous manner.
+func (b *Bridge) ListenAndServe() (conerr error) {
+	const commandPort uint32 = 0x40000000
+
+	var err error
+	b.commandConn, err = b.Transport.Dial(commandPort)
+	if err != nil {
+		return errors.Wrap(err, "bridge: failed creating the command Connection")
+	}
+	logrus.Info("bridge: successfully connected to the HCS via HyperV_Socket\n")
+
+	requestChan := make(chan *Request)
```

Comment:
Don't you want buffered channels for the request / response? The send will block until the other goroutine reads from it, so it's still serialized a bit.